sympolyc/src/expr.rs

```diff
@@ -0,0 +1,241 @@
+
+    fn add(self, other: MVPoly<F, N, D>) -> MVPoly<F, N, D> {
+        let mut result = MVPoly::new();
+        for i in 0..self.coeff.len() {
+            result.coeff[i] = self.coeff[i] + other.coeff[i];
+        }
+        result
+    }
+}
+
+impl<F: PrimeField, const N: usize, const D: usize> Add<&MVPoly<F, N, D>> for &MVPoly<F, N, D> {
+    type Output = MVPoly<F, N, D>;
+
+    fn add(self, other: &MVPoly<F, N, D>) -> MVPoly<F, N, D> {
+        let mut result = MVPoly::new();
+        for i in 0..self.coeff.len() {
+            result.coeff[i] = self.coeff[i] + other.coeff[i];
+        }
+        result
+    }
+}
+
+// Multiplication
+impl<F: PrimeField, const N: usize, const D: usize> Mul<MVPoly<F, N, D>> for MVPoly<F, N, D> {
+    type Output = Self;
+
+    fn mul(self, other: Self) -> Self {
+        let mut acc = HashMap::new();
+        let mut prime_gen = PrimeNumberGenerator::new();
+        let mut result = vec![];
+        (0..self.coeff.len()).for_each(|i| {
+            let mut sum = F::zero();
+            let normalized_index = self.normalized_indices[i];
+            let two_factors_decomposition =
+                compute_all_two_factors_decomposition(normalized_index, &mut acc, &mut prime_gen);
+            two_factors_decomposition.iter().for_each(|(a, b)| {
+                // FIXME: we should keep the inverse normalized indices
+                let inv_a = self
+                    .normalized_indices
+                    .iter()
+                    .position(|&x| x == *a)
+                    .unwrap();
+                let inv_b = self
+                    .normalized_indices
+                    .iter()
+                    .position(|&x| x == *b)
+                    .unwrap();
+                let a_coeff = self.coeff[inv_a];
+                let b_coeff = other.coeff[inv_b];
+                let product = a_coeff * b_coeff;
+                sum += product;
+            });
+            result.push(sum);
+        });
+        Self::from_coeffs(result)
+    }
+}
+
+impl<F: PrimeField, const N: usize, const D: usize> PartialEq for MVPoly<F, N, D> {
+    fn eq(&self, other: &Self) -> bool {
+        self.coeff == other.coeff
+    }
+}
+
+impl<F: PrimeField, const N: usize, const D: usize> Eq for MVPoly<F, N, D> {}
+
+impl<F: PrimeField, const N: usize, const D: usize> Debug for MVPoly<F, N, D> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
+        let mut prime_gen = PrimeNumberGenerator::new();
+        self.coeff.iter().enumerate().for_each(|(i, c)| {
+            let normalized_idx = self.normalized_indices[i];
+            if normalized_idx == 1 {
+                write!(f, "{}", c.to_biguint()).unwrap();
+            } else {
+                let prime_decomposition = naive_prime_factors(normalized_idx, &mut prime_gen);
+                write!(f, "{}", c.to_biguint()).unwrap();
+                prime_decomposition.iter().for_each(|(p, d)| {
+                    // FIXME: not correct
```

Comment:
Should this be committed?

Reply:
It is a debug method. I will fix it in a follow-up PR, if it is fine for you.